### PR TITLE
Allow inputs to be complex view tensors for Inductor CUDA graph

### LINKED ISSUE #1620: torch.split / torch.chunk cause "too complex strides" for Inductor + CUDA graph
When a model uses output of `torch.split` / `torch.chunk` ops as input, or has those ops internally, `torch._debug_has_internal_overlap` can complain about the size + strides of those tensors being too hard to determine whether memory overlap exists, preventing enablement of CUDA graph:

## Example 1: using output of `torch.split` as input to graph
```python
all_embs = torch.randn(8, 101168)
emb_split = [98400, 340, 40, 328, 380, 1320, 360]
split_emb = torch.split(all_embs, emb_split, dim=1)
```
`split_emb` now has tensors of size and stride:
```
torch.Size([8, 98400])
(101168, 1)
torch.Size([8, 340])
(101168, 1)
torch.Size([8, 40])
(101168, 1)
torch.Size([8, 328])
(101168, 1)
torch.Size([8, 380])
(101168, 1)
torch.Size([8, 1320])
(101168, 1)
torch.Size([8, 360])
(101168, 1)
```
Passing `split_emb` as input to "Inductor + CUDA graph" backend will cause CUDA graph to be disabled.

## Example 2: using `torch.chunk` within a graph
```python
_2384 = torch.randn(8, 4096)
chunk_list = torch.chunk(_2384, 16, dim=1)
```
`chunk_list` now has 16 tensors of same size and stride:
```
torch.Size([8, 256])
(4096, 1)
```
If we do this `torch.chunk` op within the graph and apply "Inductor + CUDA graph" backend, CUDA graph will be disabled.

cc. @ngimel

### PR DESCRIPTION
**Pull request by yf225**
Fixes https://github.com/pytorch/torchdynamo/issues/1620.

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire @jansel

### CODE REVIEW

File: torch/_inductor/utils.py [hunk 1]
```diff
@@ -231,6 +231,27 @@ def has_incompatible_cudagraph_ops(gm):
...
+    # Use existing tensor as-is for input tensors tracked in `known_static_input_idxs`.
+    for idx, x in enumerate(inps):
+        if idx not in known_static_input_idxs:
+            src_storage = x.storage()
+            # If common buffer is not created for this storage, create it; otherwise retrieve it.
+            if src_storage.data_ptr() not in dst_buffers:
+                dst_buffers[src_storage.data_ptr()] = torch.empty(
+                    src_storage.size(), dtype=src_storage.dtype, device=src_storage.device)
+            dst_buffer = dst_buffers[src_storage.data_ptr()]
+            storage_offset = (x.data_ptr() - x.storage().data_ptr()) // x.element_size()
```
Comment (jansel) [inline torch/_inductor/utils.py:247]: For Triton we require all inputs to be aligned to 16 bytes. (`x.data_ptr()%16==0`)

The previous version of static inputs had the side effect of also aligning the input, because `torch.zeros()` always returns aligned data.  This new version will preserve unaligned inputs, which will break things if Triton vectorizes a load from that buffer.

There is other code in this file for handling this in the non-cudagraphs case.

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_inductor/utils.py [hunk 2]
```diff
@@ -231,6 +231,27 @@ def has_incompatible_cudagraph_ops(gm):
...
+            static_inp = torch.as_strided(dst_buffer, x.size(), x.stride(), storage_offset)
+            static_inps.append(static_inp)
+        else:
+            static_inps.append(x)
```
Comment (soumith) [inline torch/_inductor/utils.py:251]: `x.detach()` now, potentially, if https://github.com/pytorch/pytorch/pull/87060 lands as-is

File: torch/_inductor/utils.py (hunk 1 above)
Comment (yf225) [reply torch/_inductor/utils.py:250]: Thanks - would this work?
```python
new_dst_buffer = torch.zeros(src_storage.size(), dtype=src_storage.dtype, device=src_storage.device)
assert new_dst_buffer.data_ptr() % 16 == 0
```

File: torch/_inductor/utils.py (hunk 2 above)
Comment (yf225) [reply torch/_inductor/utils.py]: Thanks - updated and will wait for that PR to land first

File: torch/_inductor/utils.py (hunk 1 above)
Comment (jansel) [reply torch/_inductor/utils.py:253]: The test case to check would be:
```
x, = create_static_inputs([torch.zeros(128)[1:]], {})
assert x.data_ptr() % 16 == 0
```

File: torch/_inductor/utils.py [hunk 3]
```diff
@@ -231,6 +231,35 @@ def has_incompatible_cudagraph_ops(gm):
...
+    dst_buffers = {}
+    # Create new shared storage for input tensors that have the same underlying storage.
+    # Use existing tensor as-is for input tensors tracked in `known_static_input_idxs`.
+    for idx, x in enumerate(inps):
+        if idx not in known_static_input_idxs:
+            src_storage = x.storage()
+            # If common buffer is not created for this storage, create it; otherwise retrieve it.
+            if src_storage.data_ptr() not in dst_buffers:
+                new_dst_buffer = torch.zeros(
+                    src_storage.size(),
```
Comment (ezyang) [inline torch/_inductor/utils.py:245]: Please don't do this. See my comments on the issue.
